Networking/Networking/Remote/BlazeRemote.swift

```diff
@@ -34,6 +34,29 @@ public protocol BlazeRemoteProtocol {
     ///    - query: Keyword to search for locations. Requires a minimum of 3 characters, or else will return error.
     ///
     func fetchTargetLocations(for siteID: Int64, query: String) async throws -> [BlazeTargetLocation]
+
+
+    /// Fetches forecasted campaign impressions.
+    /// - Parameters:
+    ///    - siteID: WPCom ID for the site to create ads campaigns.
+    ///    - startDate: Start date of the campaign.
+    ///    - endDate: End date of the campaign.
+    ///    - formattedTotalBudget: Formatted string of total budget of the campaign.
+    ///    - targetLocations: Target locations of the campaign. Optional.
+    ///    - targetLanguages: Target languages of the campaign. Optional.
+    ///    - targetDevices: Target devices of the campaign. Optional.
+    ///    - targetTopics: Target topics of the campaign. Optional.
+    ///
+    func fetchForecastedImpressions(
+        for siteID: Int64,
+        from startDate: Date,
+        to endDate: Date,
+        formattedTotalBudget: String,
+        targetLocationIds: [Int64],
+        targetLanguageIds: [String],
+        targetDeviceIds: [String],
+        targetTopicIds: [String]
```

Comment:
Suggestion: Instead of adding all these parameters, please create a new type `BlazeForecastedImpressionInput` with the properties and make it conform to `Encodable`. 

Then when building the request params, you can use [`toDictionary`](https://github.com/woocommerce/woocommerce-ios/blob/34a91a4e05a4e9a1fb77645fb49ad2b9666d944c/Networking/Networking/Extensions/Encodable%2BSerialization.swift#L16) to transform the object to a dictionary. This will make code more readable.

Reply:
Thanks for the feedback! The structure of the JSON request is a bit non-straightforward with "targetings" as the main key that contains these Id sub-keys.

My attempt to make an `Encodable` input that works with `.toDictionary()` looks like this, which to me looks a bit verbose:

```
public extension BlazeRemote {
    struct BlazeForecastedImpressionInput: Encodable {
        let startDate: Date
        let endDate: Date
        let formattedTotalBudget: String
        private let targetLocations: [Int64]?
        private let targetLanguages: [String]?
        private let targetDevices: [String]?
        private let targetTopics: [String]?

        private enum CodingKeys: String, CodingKey {
            case startDate = "start_date"
            case endDate = "end_date"
            case totalBudget = "total_budget"
            case targetings
        }

        private enum TargetingsKeys: String, CodingKey {
            case targetLocations = "locations"
            case targetLanguages = "languages"
            case targetDevices = "devices"
            case targetTopics = "page_topics"
        }


        public func encode(to encoder: Encoder) throws {
            var container = encoder.container(keyedBy: CodingKeys.self)

            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "yyyy-MM-dd"

            let startDateString = dateFormatter.string(from: startDate)
            let endDateString = dateFormatter.string(from: endDate)

            try container.encode(startDateString, forKey: .startDate)
            try container.encode(endDateString, forKey: .endDate)
            try container.encode(formattedTotalBudget, forKey: .totalBudget)

            var targetingsContainer = container.nestedContainer(keyedBy: TargetingsKeys.self, forKey: .targetings)
            try targetingsContainer.encode(targetLocations, forKey: .targetLocations)
            try targetingsContainer.encode(targetLanguages, forKey: .targetLanguages)
            try targetingsContainer.encode(targetDevices, forKey: .targetDevices)
            try targetingsContainer.encode(targetTopics, forKey: .targetTopics)
        }
    }
}
```

Is this what you had in mind?

Reply:
A few comments I have for the above snippet:
- We should move the struct `BlazeForecastedImpressionInput` outside as a standalone type instead of an extension of `BlazeRemote`. This will make it simpler to access from outside of the Networking module. Specifically, we'll need to create the input on the UI module and send it to Networking through `BlazeAction`.
- `BlazeForecastedImpressionInput` needs to have a public initializer to create the input from outside of the Networking module. 
- To avoid writing the custom implementation for the struct, we can add a new struct for the target options and make that struct encodable, too:

```Swift
public struct BlazeForecastedImpressionInput: Encodable {
    let startDate: Date
    let endDate: Date
    let formattedTotalBudget: String
    let targetings: BlazeTargetOptions

    public init(...) {}

    private enum CodingKeys: String, CodingKey {
        case startDate = "start_date"
        case endDate = "end_date"
        case totalBudget = "total_budget"
        case targetings
    }
}

public struct BlazeTargetOptions: Encodable {

    let targetLocations: [Int64]?
    let targetLanguages: [String]?
    let targetDevices: [String]?
    let targetTopics: [String]?

    public init(...) {}

    private enum TargetingsKeys: String, CodingKey {
        case targetLocations = "locations"
        case targetLanguages = "languages"
        case targetDevices = "devices"
        case targetTopics = "page_topics"
    }
}
```

Reply:
As for the date formatter for the encoder, you can set it through [`toDictionary`](https://github.com/woocommerce/woocommerce-ios/blob/34a91a4e05a4e9a1fb77645fb49ad2b9666d944c/Networking/Networking/Extensions/Encodable%2BSerialization.swift#L16) if the default value doesn't work for you.

Reply:
You can also send `convertToSnakeCase` to the `keyEncodingStrategy` of `toDictionary` to update the keys for `startDate` and `endDate` when encoding the struct.

Reply:
Thanks for the feedback, this is handled in https://github.com/woocommerce/woocommerce-ios/pull/11573/commits/02fe4ff9c1f9a48d572a535b58306020bfd2cd24